iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/vector.inl

```diff
@@ -186,10 +186,13 @@ inline bool vector<T, Capacity>::emplace(const uint64_t position, Targs&&... arg
     {
         return emplace_back(std::forward<Targs>(args)...);
     }
-
+    // AXIVION Next Construct AutosarC++19_03-A5.2.4 : Type-safety ensured by template parameter
+    // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast)
     emplace_back(std::move(*reinterpret_cast<T*>(m_data[m_size - 1U])));
     for (uint64_t i = m_size - 1U; i > position; --i)
     {
+        // AXIVION Next Construct AutosarC++19_03-A5.2.4 : Type-safety ensured by template parameter
+        // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast)
         reinterpret_cast<T*>(m_data)[i] = std::move(reinterpret_cast<T*>(m_data)[i - 1U]);
     }
 
```

Comment:
But I think @elBoberido agrees that we have in line 199 an actual bug.

The element at `position` was moved in the for loop in line 196. Therefore `position` does contain a valid but moved element. We have to call the destructor first before creating a new element with placement new in line 199.

So we have to add `at(position).~T()` here otherwise we override a valid element without calling its destructor first.

Here an example why this is important: the `smart_lock`. All constructors of the smart lock are creating a mutex and remove this mutex on destruction. If a `smart_lock` is moved the mutex is never moved, it is only locked and unlocked to secure the data but never moved.
So a moved `smart_lock` contains still a `mutex` resource which has to be cleaned up by the destructor otherwise we have a `mutex` leak. It is important that the `smart_lock` contains a valid `mutex` resource since this is always allowed in c++
```cxx
MyType a;
MyType b = std::move(a);
a = std::move(anotherVariable); // I can always move/copy assign another variable to a moved variable
```

Reply:
@elfenpiff Thanks for pointing that out. I have troubles to understand your example. In my view `std::move` is just passing the ownership and not ending the lifetime of an object. Hence, if someone would store a `smart_lock` in `cxx::vector` (not sure yet if that makes sense at all :D) it would still exist and the d'tor should not be called! 


```
Before
+----+----+----+----+
| 1  | 2  | 3  |    |   m_size = 3
+----+----+----+----+

vec.emplace(2, 10)

After
+----+----+----+----+
| 1  | 10 | 2  | 3  |   m_size = 4
+----+----+----+----+
```

Element `2` is still alive, so why should the d'tor be called?

Reply:
@mossmaurice In line 199 you perform a placement new, this overrides an object without calling the destructor.

The lifetime of an object always has to start with the constructor and always has to end with the destructor. When one performs another placement new on a memory position were already an object is living one breaks this essential contract. The destructor must be called before the placement new.

Reply:
@elfenpiff In the above example this rule is not violated. Once `2` get removed from the vector the d'tor will be called. Overriding a move-from object is ok. Think about the empty-house-without-furniture analogy. We can construct and new house on a empty house.

Reply:
Ok, got it now. Although, `std::move` means passing the ownership of ressources according to RAII, there can be other things in the object that are not moved. This needs to be destroyed of course. :see_no_evil: Will write a test case and fix it.